File: CIT/julia/src/Statistics.jl
module KernelStatistics

using Statistics

struct KernelSummary
    mean::Vector{Float64}
    lower::Vector{Float64}
    upper::Vector{Float64}
    susceptibility::Float64
    positive_susceptibility::Float64
    threshold::Float64
    support_95::Float64
    support_997::Float64
end

trapz(t, y) = sum((y[1:end-1] .+ y[2:end]) .* diff(t)) / 2

function summarize_kernel(times::Vector{Float64}, blocks::Matrix{Float64}; alpha=0.05)
    alpha == 0.05 || throw(ArgumentError("the stdlib implementation currently supports alpha=0.05"))
    μ = vec(mean(blocks; dims=2))
    se = vec(std(blocks; dims=2, corrected=true)) ./ sqrt(size(blocks, 2))
    z = 1.959963984540054
    lower, upper = μ .- z .* se, μ .+ z .* se
    susceptibility = trapz(times, μ)
    positive = max.(μ, 0.0)
    positive_susceptibility = trapz(times, positive)
    positive_susceptibility > 0 || throw(ArgumentError("positive susceptibility must be non-zero"))
    increments = vcat(0.0, (positive[1:end-1] .+ positive[2:end]) .* diff(times) ./ 2)
    share = cumsum(increments) ./ sum(increments)
    i95 = findfirst(>=(0.95), share)
    i997 = findfirst(>=(0.997), share)
    KernelSummary(μ, lower, upper, susceptibility, positive_susceptibility,
                  inv(positive_susceptibility), times[i95], times[i997])
end

laplace_transform(times, kernel, args) = [trapz(times, exp.(-x .* times) .* kernel) for x in args]

end

File: CIT/python/src/cit/statistics.py
from __future__ import annotations

from dataclasses import dataclass
import numpy as np
from scipy.stats import norm

@dataclass(frozen=True)
class KernelSummary:
    mean: np.ndarray
    lower: np.ndarray
    upper: np.ndarray
    susceptibility: float
    positive_susceptibility: float
    threshold: float
    positive_threshold: float
    support_95: float
    support_997: float

def summarize_kernel(times: np.ndarray, blocks: np.ndarray) -> KernelSummary:
    mean = blocks.mean(axis=1)
    se = blocks.std(axis=1, ddof=1) / np.sqrt(blocks.shape[1])
    z = norm.ppf(0.975)
    lower, upper = mean - z * se, mean + z * se
    susceptibility = float(np.trapezoid(mean, times))
    if susceptibility <= 0.0:
        raise ValueError("raw susceptibility must be positive")
    positive = np.maximum(mean, 0.0)
    positive_susceptibility = float(np.trapezoid(positive, times))
    if positive_susceptibility <= 0.0:
        raise ValueError("positive susceptibility must be non-zero")
    increments = np.r_[0.0, 0.5 * (positive[1:] + positive[:-1]) * np.diff(times)]
    share = np.cumsum(increments) / increments.sum()
    return KernelSummary(
        mean, lower, upper,
        susceptibility, positive_susceptibility,
        1.0 / susceptibility,
        1.0 / positive_susceptibility,
        float(times[np.argmax(share >= 0.95)]),
        float(times[np.argmax(share >= 0.997)]),
    )

def laplace_transform(times: np.ndarray, kernel: np.ndarray, arguments: np.ndarray) -> np.ndarray:
    return np.array([np.trapezoid(np.exp(-x * times) * kernel, times) for x in np.atleast_1d(arguments)])
